File: findLargestValueInRow.java
//https://leetcode.com/problems/find-largest-value-in-each-tree-row/
// Time Complexity :O(n)  
// Space Complexity :O(n/2) 
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :no


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        ArrayList<Integer> result=new ArrayList<>();
        if(root==null) return result;
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int size=q.size(); 
            int max=Integer.MIN_VALUE;
            ArrayList<Integer> temp=new ArrayList<>();
            for(int i=0;i<size;i++)
            {
                TreeNode curr= q.poll();
                if(curr.val>max)
                    max=curr.val;
                if(curr.left!=null) 
                {
                 
                    q.add(curr.left);
                }
                if(curr.right!=null) 
                {
                    
                    q.add(curr.right);
                }
              
            result.add(max);
        }
        return result;
    }
}

File: Find_largest_value_in_row.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# TC: O(N)
# SC: O(H)

class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        q=deque()
        q.append(root)

        res=[]
        while q:
            size1=len(q)
            max1=-999999999999999999
            for i in range(0,size1):
                h=q.popleft()
                max1=max(max1,h.val)

                if h.left:
                    q.append(h.left)
                
                if h.right:
                    q.append(h.right)
            res.append(max1)

        return res




        